### PR TITLE
Concurrency

### PR DESCRIPTION
**Pull request by nrwiersma**

### CODE REVIEW

**Comment by nrwiersma** [comment]
Still need to fix the tests, but review can begin none the less

File: pipe.go [hunk 1]
```diff
@@ -6,58 +6,58 @@ import (
...
 
-// SetNode sets the topology node that is being processed.
+// Queue gets the queued Messages for each Node.
 //
-// The is only needed by the task and should not be used
-// directly. Doing so can have some unexpected results.
-func (p *ProcessorPipe) SetNode(n Node) {
-	p.currentNode = n
+// Reading the node message queue will reset the queue.
+func (p *ProcessorPipe) Queue() []NodeMessage {
```
Comment (michalkurzeja) [inline pipe.go]: Do we really need the `Queue` method exposed, especially if it has such a dangerous side-effect?
Also, because of this side-effect, the method should be called `PopQueue()` or `ConsumeQueue()`

File: task.go [hunk 2]
```diff
@@ -2,69 +2,137 @@ package streams
...
+			}
+
+			if msg.Empty() {
+				continue
+			}
+
+			stats.Timing(msg.Ctx, "node.latency", time.Since(start), 1.0, "name", node.Name())
+			stats.Inc(msg.Ctx, "node.throughput", 1, 1.0, "name", node.Name())
+
+			nodeMsgs, _ := node.Process(msg)
```
Comment (michalkurzeja) [inline task.go]: Why are you ignoring the error here?

File: pipe.go (hunk 1 above)
Comment (nrwiersma) [reply pipe.go]: Generally I agree, but for this purpose, we unfortunately cannot. I have plans to get rid of this in a future release.
